File: process/datastore/constructions/composition.cpp
#include "construction.h"
#include "../datastore.h"
#include "../util/math.h"

class c_composition : public construction {
public:
	c_composition() : construction( "composition", 26 ) {
	}

	void process( array * base ) {
		// process as main array
		for ( int v = base->t; v <= MAXV; v++ ) {
			array * subarray = getBestArray( base->v, v, base->t );

			if ( subarray != NULL ) {
				array * arr = new array();
				arr->N = base->N * subarray->N;
				arr->k = base->k;
				arr->v = v;
				arr->t = base->t;

				arr->type = 'C';
				arr->source = id;

				arr->ingredients.push_back( base );
				arr->ingredients.push_back( subarray );

				insertArray( arr );
			}
		}

		// process as subarray
		array * prev = getPreviousArray( base );		

		if ( prev != NULL && (prev->k >= MAXV ))
			return;

		fullSortTableType arrays;
		query( arrays, base->t, base->t, prev != NULL ? prev->v + 1 : base->t, base->k, 0, -1 );

		fullSortTableType::iterator superIt;
		for ( superIt = arrays.begin(); superIt != arrays.end(); superIt++ ) {		
			array * superarray = (*superIt);

			array * arr = new array();
			arr->N = base->N * superarray->N;
			arr->k = superarray->k;
			arr->v = base->v;
			arr->t = base->t;

			arr->type = 'C';
			arr->source = id;
			
			arr->ingredients.push_back( superarray );
			arr->ingredients.push_back( base );

			insertArray( arr, &arrays );
		}

	}
};

// declare the construction object
construction * composition = new c_composition();

File: process/constructions/composition.pl
#!/usr/bin/perl

my $name = 'composition';
my $id;

$id = get_construction_id( $name, 7, 2 );

$constructions{ $name } = {};

$constructions{ $name }->{'process_array'} =
	sub {
		my $array = shift;

		# process as main array
		my $v;
		for ( $v = $array->{'t'}; $v <= $MAXV; $v++ ) {
			$subarray = get_best_array( $array->{'v'}, $v, $array->{'t'} );

			if ( $subarray ) {
				$new_array = new_constructed_array( $array->{'N'} * $subarray->{'N'},
		    	                                    $array->{'k'},
												    $v,
												    $array->{'t'} );

				$new_array->{'construction'} = $id;
				$new_array->{'ingredients'} = [ $array, $subarray ];

				persist_constructed_array( $new_array );
			}
		}

		# process as subarray
		my $prev = get_max_k_array( 'k < ' . $array->{'k'} . ' AND t = ' . $array->{'t'} . ' AND v = ' . $array->{'v'} );

		return if ( $prev && ($prev->{'k'} >= $MAXV ));

		my $condition = 't = ' . $array->{'t'} . ' AND v <= ' . $array->{'k'};
		if ( $prev ) {
			$condition .= ' AND v > ' . $prev->{'k'};
		}

		my $arrays = get_arrays( $condition );
		my $superarray;
		foreach $superarray (@$arrays) {
			my $new_array = new_constructed_array( $array->{'N'} * $superarray->{'N'},
				                                   $superarray->{'k'},
					   							   $array->{'v'},
												   $array->{'t'} );
			$new_array->{'construction'} = $id;
			$new_array->{'ingredients'} = [ $superarray, $array ];

			persist_constructed_array( $new_array );
		}

	};




1;
